### DISCUSSION TITLE
How I make python subprocess Popen work with quarto command?

### DISCUSSION THREAD

**Discussion by wagnerc4** [quarto-dev/quarto-cli#8743] [Q&A]
### Description

I try to send a input string to quarto and try to receive a output string.

    from subprocess import Popen, PIPE
    
    p = Popen(["/opt/quarto/bin/quarto", "render"], stdin=PIPE, stdout=PIPE, stderr=PIPE, encoding="UTF-8")
    rs, err = p.communicate(input="# Hello Wold")
    print(err)

Please help!
Thanks!

**Comment by mcanouil** [comment]
I don't think Quarto can read from standard input.
Before doing things in another language, I suggest you try directly using Quarto CLI in your terminal.

**Comment by wagnerc4** [comment]
quarto is a very nice program, but if I can't integrate it into my programs I can't use it... I think it's a mistake that it can only be used from the terminal.

**Reply by cscheid** [reply]
> I think it's a mistake that it can only be used from the terminal.

You can easily use quarto from other programs. It's just that you can't use standard input right now. Create a temporary file and pass that as the parameter.

**Reply by wagnerc4** [reply]
I don't like that solution... this adds unnecessary processes to the program, also reading the output files and then deleting them.

**Comment by wagnerc4** [comment]
Making this great program is the hardest part! 

Making it easy to integrate is nothing!

**Reply by mcanouil** [reply]
Please be mindful of your tone.
This type of comments is not helpful or useful, see <https://github.com/quarto-dev/quarto-cli?tab=coc-ov-file#readme>.

If you don't like the current behaviour, it's fine, but keep in mind, there is a tone of work done and more to come, while time is limited.

> Making it easy to integrate is nothing!

Note that pull requests are welcome. If that is really "nothing" as you say, feel free to contribute.

**Reply by wagnerc4** [reply]
Sorry, my intention was not to make you angry!  I think it was misinterpreted, with nothing I meant that it was small.

**Reply by wagnerc4** [reply]
In any case, it is important to give priority to this function ;)

**Reply by mcanouil** [reply]
> angry

I'm definitely not.
It was simply a reminder of the code of conduct motivated by the following two replies which does not come nice together in addition of not bringing more to the issue.
Explaining your use case and why using standard input would be important to add compared to all other features would be better/valuable for instance.

> I don't like that solution... this adds unnecessary processes to the program, also reading the output files and then deleting them.

> Making this great program is the hardest part!
> 
> Making it easy to integrate is nothing!

Note: I am marking all these as off topic.

**Comment by wagnerc4** [comment]
Most people don't use the terminal.  Then we need to write programs for end users (web or desktop applications), therefore it needs to be easy to integrate.

**Reply by cscheid** [reply]
We've heard your request, we've responded to it, and it's recorded. You don't need to continue posting now.
